h2/src/main/org/h2/engine/Session.java

```diff
@@ -994,7 +1001,8 @@ private void unlockAll() {
             DbException.throwInternalError();
         }
         if (!locks.isEmpty()) {
-            for (Table t : locks) {
+            Table[] array = locks.toArray(new Table[0]);
+            for (Table t : array) {
                 t.unlock(this);
```

Comment:
If `locks` is modified by multiple threads concurrently, `t` may be `null` here, `ArrayList.toArray()` and `ArrayList.toArray(T[])` aren't fully thread-safe. They can't fail, but they can return an array with leading or trailing nulls or duplicate elements.

Reply:
I know, it’s not, but with this PR changes, chance of this being executed concurrently should be “win the lottery” type of thing. Maybe catch ConcurrentModificationException and retry will be a safer option?

Reply:
AFAIK, `ConcurrentModificationException` was added to thread-unsafe classes to notify developers that use them that their code is broken. It is not reliable.

From my point of view a simple null check for `t` is required here to avoid NPE if list is cleared by concurrent thread during execution of `toArray()`.

Reply:
If two `unlock()` calls for the same object are allowed, null check should be enough. If not, a more reliable solution is required with thread-safe collection or with lock around it.